src/frequenz/channels/util/_select.py

```diff
@@ -117,82 +80,41 @@ async def stop(self) -> None:
         await asyncio.gather(*self._pending, return_exceptions=True)
         self._pending = set()
 
-    async def ready(self) -> bool:
+    async def ready(self) -> AsyncIterator[Set[Receiver[Any]]]:
```

Comment:
By returning `Set`s from the iterator,  we end up losing two features present in the current implementation, that protect users from hard-to-debug data loss:

   1. If a user writes `elif` instead of `if` for a condition check,  the current implementation ensures that there will be no data loss,  but the proposed design doesn't seem to have that guarantee.  This could lead to hard-to-detect/hard-to-reproduce bugs,  in most of our (relatively slow) streams,  because there will be data loss only on those rare cases where multiple receivers become ready in the same `Select` iteration.
    
   2. If someone removes an `if` check, but forgets to remove the corresponding receiver from the `Select` call (or if they add it to `Select` and forget to check for it),  the current implementation prints warnings,  whereas these get silently ignored in the proposed design.
   
   Both these features are essential in my opinion,  and they can be addressed by returning a custom type that holds the `set/list` objects instead.

Reply:
I don't understand 1. This implementation is heavily focus on preventing data loss. It is the other way around, in the previous implementation if you write a `elif` then the message will be explicitly dropped for not handled messages, while this implementation will never drop a message unless `consume()` is called, which can be only be done by the user. Worse case in this implementation, if you use `elif` one of your receivers could starve if a receiver above it has messages all the time, otherwise eventually the `elif` will be hit and the `receiver.consume()` will get the message.

Or maybe you are talking about the current implementation dropping messages working as a mechanism to make the receiver's queue move even in the event of bugs? In this implementation if messages are never processed the queue will fill up and messages will be dropped on the other end. But again, this will only happen if you have a bug **AND** you have a receiver that is receiving messages on every loop.

About 2., you would have to remove the `if` check, not remove the receiver from the `Select` **BUT** keep the receiver around for this to be a problem, because if you remove the receiver entirely, then you'll get an error about the undefined symbol when trying to use it in `Select`. It also looks to me an obscure enough case to design the API around preventing that.

For me adding an abstraction on top of the naked receiver just makes the API (and implementation) more complicated to only cover for what IMHO is an obscure usage bug.

Reply:
> It is the other way around, in the previous implementation if you write a elif then the message will be explicitly dropped for not handled messages

I don't think this is correct.  Why do you think that this is the case?  Have you tested this/seen this behaviour somewhere or are you just guessing?

> Worse case in this implementation, if you use elif one of your receivers could starve if a receiver above it has messages all the time, otherwise eventually the elif will be hit and the receiver.consume() will get the message.

This sounds very bad for a general purpose `Select` implementation.  I'd say this alone is a good enough reason for introducing a custom response type.

> It also looks to me an obscure enough case to design the API around preventing that.

it is not obscure at all,  it can happen all the time, especially if this is well used,  and this is the reason we have things like exhaustive matching these days,  like in the new python match statement.

For example,  imagine writing a select with 6 receivers,  and one of them is a `config` update channel and you forget to handle it in the select block.  Then when there are config changes,  you'll get a notification only when the buffer fills up,  and the buffer has a default size of 50.

There can be several similar scenarios.  So I think this definitely needs to be handled.

Reply:
> It is the other way around, in the previous implementation if you write a elif then the message will be explicitly dropped for not handled messages
<details>

<summary>This is not accurate.  Just tested:</summary>

``` python
import asyncio

from frequenz.channels import Broadcast
from frequenz.channels.util import Select


async def test():
    chan1 = Broadcast("")
    chan2 = Broadcast("")

    select = Select(c1=chan1.new_receiver(), c2=chan2.new_receiver())

    for ii in range(10):
        await chan1.new_sender().send(ii)
        await chan2.new_sender().send(ii + 10)

    await chan1.close()
    await chan2.close()

    while await select.ready():
        print(f"ready items: {select._ready_count}")
        if val := select.c1:
            print(f"Chan1 says: {val.inner}")
        elif val := select.c2:
            print(f"Chan2 says: {val.inner}")

        if select._ready_count == 0:
            print()


asyncio.run(test())

```
</details>

Reply:
> I don't think this is correct. Why do you think that this is the case? Have you tested this/seen this behaviour somewhere or are you just guessing?

No, I guess my memory is only faulty. We had this exact same conversation a while ago and you also explained that `elif` was OK but I guess for some reason that didn't stuck with me and I kept the misconception in my mind all the time. It's all in:

* https://github.com/frequenz-floss/frequenz-channels-python/issues/27

You said there that using `break` was problematic, but now I don't understand why can that be the case, unless you break the loop before processing any messages at all, as the current select will drop messages if no messages were consumed at all in one loop iteration (I still don't understand why is that but that's a different topic), but doing:

```py
    while await select.ready():
        print(f"ready items: {select._ready_count}")
        if val := select.c1:
            print(f"Chan1 says: {val.inner}")
            break
        if val := select.c2:
            print(f"Chan2 says: {val.inner}")
```

Should be fine too because if the `break` was reached, it was because at least one receiver message was consumed. I'm not sure why we thought that was a problem in that issue anymore.

What should drop messages is if you forget to handle one case inside the loop, and you only receive a message for that receiver.

OK, so let's forget about this for the current implementation. The point is the current implementation doesn't drop message under any circumstances. :D

Reply:
>> Worse case in this implementation, if you use elif one of your receivers could starve if a receiver above it has messages all the time, otherwise eventually the elif will be hit and the receiver.consume() will get the message.
> This sounds very bad for a general purpose Select implementation. I'd say this alone is a good enough reason for introducing a custom response type.

This is true for the current implementation too. I agree it would be good to prevent it, but I would leave it for another PR as it is a different enhancement.

Reply:
> For example, imagine writing a select with 6 receivers, and one of them is a config update channel and you forget to handle it in the select block. Then when there are config changes, you'll get a notification only when the buffer fills up, and the buffer has a default size of 50.

OK, so you are not saying the case where you remove some receiver but when you are writing a select loop for the first time and forgot to handle one case. Fair enough then. I agree it would be idea to leverage `match` here then, if there is a way to let mypy tell you if a case is not being properly handled at "compile time" instead of runtime, that would be awesome.

I'll try to look into this option and if not we keep the runtime check.

Reply:
> Should be fine too because if the break was reached, it was because at least one receiver message was consumed. I'm not sure why we thought that was a problem in that issue anymore.

Just for the records, this was an issue and it was fixed:

* https://github.com/frequenz-floss/frequenz-channels-python/pull/42

Reply:
> You said there that using break was problematic, but now I don't understand why can that be the case, unless you break the loop before processing any messages at all,

yes,  break was problematic,  and we fixed it by splitting `receive` into `ready` and `consume`.

> as the current select will drop messages if no messages were consumed at all in one loop iteration (I still don't understand why is that but that's a different topic)

This allows us to log a warning about an unread message as soon as we know it is unread,  for each missed message,  because that is a clear bug in user code.

>>> Worse case in this implementation, if you use elif one of your receivers could starve if a receiver above it has messages all the time, otherwise eventually the elif will be hit and the receiver.consume() will get the message.
>>
>> This sounds very bad for a general purpose Select implementation. I'd say this alone is a good enough reason for introducing a custom response type.
>
> This is true for the current implementation too. I agree it would be good to prevent it, but I would leave it for another PR as it is a different enhancement.

no,  the current implementation doesn't starve any of the receivers, whether or not `elif` is used,  as demonstrated in the example I shared above.  And it is going away in the PR.

> OK, so you are not saying the case where you remove some receiver but when you are writing a select loop for the first time and forgot to handle one case

yes, I mean there can be we can get into such a situation from either direction,  and sometimes it will be a slow but critical thing and we might not notice.

Reply:
> This allows us to log a warning about an unread message as soon as we know it is unread, for each missed message, because that is a clear bug in user code.

I'm not sure about this. Even when convoluted, you could check some condition before consuming something that is ready and decide to not consume it yet and give the loop another spin. I would rather avoid dropping messages under any circumstance. Logging is OK, as it is likely a bug, and certainly not good programming if it isn't, but dropping is going a bit too far IMHO. If the code has a bug you won't make it better by just dropping messages and continue with your life as if nothing happened.

Reply:
> no, the current implementation doesn't starve any of the receivers, whether or not elif is used, as demonstrated in the example I shared above. And this feature is going away in this PR.

Ah, right, because you don't schedule the receiver that was already consumed to receive again until all other ready receivers were consumed. I see :+1: